### DISCUSSION TITLE
Add page-based content with Nunjucks variables

### DISCUSSION THREAD

**Discussion by Brixy** [11ty/eleventy#1922] [Q&A] [answered]
Hi all!

On a page basis we want to add (optional) ‘meta data’–like page headers, page footers, `<aside />` etc.

Global Nunjucks variables [are not supported](https://github.com/11ty/eleventy/issues/685). So I came up with the code below. `meta.njk` needs to be added multiple times.

As we are planning a rather large-scale eleventy project I would like to ask you: Is this the way to go, or would you suggest another method?

Front matter for `pageHeader` and `pageFooter` is not an option because the contents are too complex.

Thank you very much in advance!

**Content file:**

```
---
title: Test page
---
{%- set pageHeader -%}
My page header 1 content.
{%- endset -%}
{% include "meta.njk" %}

{%- set pageHeader -%}
My page header 2 content.
{%- endset -%}
{% include "meta.njk" %}

<!-- page contents -->

{%- set pageFooter -%}
My page footer content.
{%- endset -%}
{% include "meta.njk" %}
```

**Partial `meta.njk`:**

```
{% if pageHeader %}
<header class="page-header">
  {{ pageHeader | safe }}
</header>
{% endif %}

{% if pageFooter %}
<footer class="page-footer">
  {{ pageFooter | safe }}
</footer>
{% endif %}
```

**Comment by pdehaan** [comment]
I think support for Nunjucks globals was added in https://github.com/11ty/eleventy/pull/1060 (which will ship w/ Eleventy v1.0). 🎉

**Comment by pdehaan** [comment, marked as answer]
I'm not sure if I follow 100%, but I don't know if that will work (although admittedly I haven't tried building it locally yet). I think my main concerns would be that `pageHeader` gets set, and then added to the page using `{% include "meta.njk" %}`. Then we overwrite the local `pageHeader` variable w/ page header 2 content and add another `<header class="page-header">{{ pageHeader | safe }}</header>` to the DOM. So far so good, but you'll have two `<header class="page-header">`s on your page. But now when you get to the footer, I _assume_ your `pageHeader` variable from earlier is still in scope, so your footer might have the following:

```html
<header class="page-header">
  My page header 2 content.
</header>

<footer class="page-footer">
  My page footer content.
</footer>
```

Although now you've made me very curious whether I guessed correctly, so I'll try building a small proof of concept locally.

---

```njk
---
title: Test page
---
{%- set pageHeader -%}
  My page header 1 content.
{%- endset -%}
{% include "meta.njk" %}

{%- set pageHeader -%}
  My page header 2 content.
{%- endset -%}
{% include "meta.njk" %}

<!-- page contents -->

{%- set pageFooter -%}
  My page footer content.
{%- endset -%}
{% include "meta.njk" %}
```

**Reply by pdehaan** [reply]
![Test_page](https://user-images.githubusercontent.com/557895/128612887-a1bbac08-0731-45fa-a5ff-3c1156ad04c8.png)

You'd either have to clear the `pageHeader` and `pageFooter` after using them `{% set pageHeader = false %}` so they don't leak into the footer area, or else create separate "metaHeader.njk" and "metaFooter.njk" include files to avoid leaking variables. Or even separate layout files for different page types; it's hard to tell without seeing the site source.

**Comment by Brixy** [comment]
Excellent! Thank you very much for you support and your estimation.

Using seperate "metaHeader.njk" and "metaFooter.njk" partials is a practical solution.

I had thought of using macros, shortcodes and and methods; but your comments encourage me to use the partial method.

Thanks again.

**Reply by pdehaan** [reply]
Yeah, definitely a few different approaches that could work&hellip;
I imagine even paired shortcodes, like:

```njk
{% pageheader %}
  My page header 1 content.
  My page header 2 content.
{% endpageheader %}
```

Or Nunjucks blocks is probably a solution:

```njk
{% block pageheader %}
  My page header 1 content.
  My page header 2 content.
{% endblock %}

{% block pagecontent %}
  <main>
	<p>Some stuff here</p>
  </main>
{% endblock %}

{% block pagefooter %}
  My page footer content.
{% endblock %}
```

**Reply by Brixy** [reply]
Yeah, a simple paired shortcode is a good alternative.

The block solution seems to be too Nunjucks-specific.

Choosing between a shortcode or Eleventy 1.0’s new global variables seems to be a 50/50 choice ;-)

Thank you, Peter.

**Reply by pdehaan** [reply]
Here was my little proof of concept on Nunjucks blocks and paired shortcodes:

```njk
---
# ./src/index.njk
title: Homepage, but with blocks!
---

{% extends "layouts/base.njk" %}

{% block pageheader %}
  {% pageheader %}
    <p>Page header 1</p>
    <p>Page header 2</p>
  {% endpageheader %}
{% endblock %}

{% block pagefooter %}
  {% pagefooter %}
    <p>Page footer 1</p>
  {% endpagefooter %}
{% endblock %}

{% block pagecontent %}
  <p>I am some content.</p>
{% endblock %}
```

```njk
<!-- ./src/_includes/layouts/base.njk -->
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>{{ title }}</title>
  <style>
    .page-header {
      background-color: salmon;
      padding: 1em;
      margin-bottom: 0.5em;
    }
    .page-footer {
      background-color: gold;
      padding: 1em;
      margin-bottom: 0.5em;
    }
  </style>
</head>
<body>

  {% block pageheader %}{% endblock %}
  
  <main>
  {% block pagecontent %}{% endblock %}
  </main>

  <aside>
  {% block pagefooter %}{% endblock %}
  </aside>

</body>
</html>
```

### OUTPUT

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Homepage, but with blocks!</title>
    <style>
      .page-header {
        background-color: salmon;
        padding: 1em;
        margin-bottom: 0.5em;
      }
      .page-footer {
        background-color: gold;
        padding: 1em;
        margin-bottom: 0.5em;
      }
    </style>
  </head>
  <body>
    <header class="page-header">
      <p>Page header 1</p>
      <p>Page header 2</p>
    </header>

    <main>
      <p>I am some content.</p>
    </main>

    <aside>
      <footer class="page-footer">
        <p>Page footer 1</p>
      </footer>
    </aside>
  </body>
</html>
```

**Reply by Brixy** [reply]
Thank you. That’s a great reference.

Apart from the technical point of view maybe you have to decide between a more content-ish approach (shortcode) or a template-ish approach (partial/block).
